x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java

```diff
@@ -319,9 +317,16 @@ private boolean assertRangeIsAlignedWithPart(Tuple<Long, Long> range) {
         return true;
     }
 
-    private int readCacheFile(FileChannel fc, long end, long position, byte[] buffer, int offset, long length) throws IOException {
+    private int readCacheFile(FileChannel fc, long end, long position, ByteBuffer b, long length) throws IOException {
         assert assertFileChannelOpen(fc);
-        int bytesRead = Channels.readFromFileChannel(fc, position, buffer, offset, Math.toIntExact(Math.min(length, end - position)));
+        int bytesRead;
+        if (end - position < b.remaining()) {
+            int l = Math.toIntExact(end - position);
+            b.put(Channels.readFromFileChannel(fc, position, l));
```

Comment:
We should try to read directly into the ByteBuffer instead of materializing a byte[]?

Reply:
If I recall correctly if the size of the data is lower that the remaining size of the byte buffer you get an EOF exception. We are only materialising to a byte[] in that case. Maybe there is a better way?